recipes/moltenvk/all/conanfile.py

```diff
@@ -0,0 +1,155 @@
+    url = "https://github.com/conan-io/conan-center-index"
+
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "with_spirv_tools": [True, False],
+        "tools": [True, False]
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+        "with_spirv_tools": True,
+        "tools": True
+    }
+
+    exports_sources = ["CMakeLists.txt", "patches/**"]
+    generators = "cmake"
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        if self.settings.compiler.get_safe("cppstd"):
+            tools.check_min_cppstd(self, 11)
+        if self.settings.os not in ["Macos", "iOS", "tvOS"]:
+            raise ConanInvalidConfiguration("MoltenVK only supported on MacOS, iOS and tvOS")
+
+    def requirements(self):
+        self.requires("cereal/1.3.0")
+        self.requires("glslang/8.13.3559")
+        self.requires("spirv-cross/{}".format(self._spirv_cross_version))
+        self.requires("vulkan-headers/{}".format(self._vulkan_headers_version))
+        if self.options.with_spirv_tools:
+            self.requires("spirv-tools/v2020.5")
+        if tools.Version(self.version) < "1.1.0":
+            raise ConanInvalidConfiguration("MoltenVK < 1.1.0 requires vulkan-portability")
+
+    @property
+    def _spirv_cross_version(self):
+        return {
+            "1.1.1": "20210115", # can't compile with spirv-cross < 20210115
+            "1.1.0": "20200917", # works with spirv-cross 20200917 only
+        }.get(self.version)
+
+    @property
+    def _vulkan_headers_version(self):
+        return {
+            "1.1.1": "1.2.162.0",
+            "1.1.0": "1.2.154.0",
+            "1.0.44": "1.2.148.0",
+            "1.0.43": "1.2.141.0",
+            "1.0.42": "1.2.141.0",
+            "1.0.41": "1.2.135.0",
+            "1.0.40": "1.2.131.1",
+            "1.0.39": "1.1.130.0",
+            "1.0.38": "1.1.126.0",
+            "1.0.37": "1.1.121.0",
+            "1.0.36": "1.1.114.0",
+            "1.0.35": "1.1.108.0",
+            "1.0.34": "1.1.106.0",
+            "1.0.33": "1.1.101.0",
+            "1.0.32": "1.1.97.0",
+            "1.0.31": "1.1.97.0",
+            "1.0.30": "1.1.92.0",
+            "1.0.29": "1.1.92.0",
+            "1.0.28": "1.1.92.0",
+            "1.0.27": "1.1.92.0",
+            "1.0.26": "1.1.85.0",
+            "1.0.25": "1.1.85.0",
+            "1.0.24": "1.1.85.0",
+            "1.0.23": "1.1.85.0",
+            "1.0.22": "1.1.82.0",
+            "1.0.21": "1.1.82.0",
+            "1.0.20": "1.1.82.0",
+            "1.0.19": "1.1.82.0",
+            "1.0.18": "1.1.82.0",
+            "1.0.17": "1.1.82.0",
+        }.get(self.version)
+
+    def package_id(self):
+        del self.info.settings.compiler.version
```

Comment:
Do not remove the `compiler.version`. Even if it only works with apple-clang 12, it is needed.

Reply:
The problem is that MoltenVK > 1.0.42 must be built with XCode >= 12, but can be consumed with XCode < 12...

Reply:
That's the use-case for `validate` + `compatible_packages`: we declare that the recipe can only be built with XCode 12, but then in the `compatible_packages` we list all the versions that can consume it (it will be the same scenario for `compiler.cppstd`).

It is the idea I tried to ilustrate here (https://github.com/conan-io/examples/pull/73), have a look at [the `conanfile.py`](https://github.com/conan-io/examples/pull/73/files#diff-546291ee317991514c70c1331f5be6c4ee383251143b212d77a639ed0be900f1), more or less:

```python
class Recipe(ConanFile):
    def configure(self):
        if self.settings.os not in ["Macos", "iOS", "tvOS"]:
            raise ConanInvalidConfiguration("MoltenVK only supported on MacOS, iOS and tvOS")

    def validate(self):
        if tools.Version(self.settings.compiler.version) < "12":
            raise ConanInvalidConfiguration("MoltenVK {} requires XCode 12 or higher at build time".format(self.version))

    def package_id(self):
        for xcode in ("9", "10", "11"):
            compatible_pkg = self.info.clone()
            compatible_pkg.settings.compiler.version = xcode
            self.compatible_packages.append(compatible_pkg)
```

The problem with removing `compiler.version` is that CCI identifies that it is the same `package-id` for all the macOS builds (9.1, 10, 11, 12), CCI removes duplicates and chooses to build with the oldest of all the compilers that are available. CCI chooses apple-clang 9.1 and then it founds the `ConanInvalidConfiguration` in the `build()` method.

Reply:
Shouldn't it be something like this?

```python
    def package_id(self):
        if tools.Version(self.settings.compiler.version) < "12":
                compatible_pkg = self.info.clone()
                compatible_pkg.settings.compiler.version = "12"
                self.compatible_packages.append(compatible_pkg)
```

Reply:
🤕  you are right regarding `package_id()` implementation.

Conan captures the `ConanInvalidConfiguration` if it is raised from the `validate()` method. And will it use in different ways depending on the command. If it `conan info`, Conan will print that the computed package-ID is invalid instead of raising the error, if it is a `conan create/build` then it will raise.

Besides the UX, we are preparing for Conan 2.0 where two-profiles would be the default. Still we don't know what is going to happen with the `build-requires`, if we will expand the full graph always (also the build-requires). If that is the case, when you are installing packages for a _host_, you really don't care (and probably don't know) the profile used to build them, the recipes in the _build_ context will probably get a default (?) empty (?) profile that will likely raise a `ConanInvalidConfiguration`.... but we will capture the exception unless you really want to use/build those binaries and Conan needs to retrieve/build the package. This is WIP, probably it will change, but it fits with the better UX experience.